### ISSUE TITLE
Deal with possibility of cycles in xml values

### ISSUE THREAD

**Issue by jclark** [ballerina-platform/ballerina-spec#410]
There is a long-standing issue with the design of the xml type on how to deal with cycles.

Since xml:Element is mutable, there is the possibility of cycles, e.g.:

```
e.setChildren(e);
```

Such a value would not be serializable.  We need to figure what to do about this. XML operations need to behave robustly in all circumstances. The best solution might be for setChildren to catch this (and panic), so that nothing else needs to deal with this, but I am not sure if this can be done efficiently.

**Comment by rdhananjaya** [comment]
I also don't see how we can efficiently detect the cycle since we will have to search the child tree. Can we make the semantic of `setChildren` so that it always set a clone of provided child tree?

**Comment by jclark** [comment]
Having `setChildren` clone is an interesting idea, but it would have some consequences that I think are undesirable:

- there would be a big difference between (a) constructing the element with a constructor , then adding children with setChildren and (b) constructing the element and specifying the children in the constructor; but they ought to be similar
- it would be too easy for an element at depth N to be cloned  N times during the course of tree construction
- part of the thinking behind the XML design is to efficiently support the scenario where you have an XML document representing an envelope wrapping a body, where the body is big relative to the envelope, and you want to efficiently put the body in a new envelope; efficiently here meaning without copying

**Comment by jclark** [comment]
On the other hand doing cycle detection is going to be quite hard especially when you consider thread-safety:  there is a possibility for multiple threads to be simultaneously mutating an XML tree, and the Ballerina runtime must deal robustly with thus (it should not crash or run out of resources )

**Comment by jclark** [comment]
Here's an idea for how to deal with this. (It is a kind of dynamic ownership typing.)

Every element _owns_ its mutable child elements. Element construction and mutation maintain the invariant that a mutable element never has more than one owner.

An XML element has a _owned flag_. If the owned flag is off, it means that this element is known not to be owned by any element (i.e. known not to be a constituent of the children of any element). 

For example, with

```
xml`<a><b/></a><a><b><c/></b></a>`
```

both the `<a>` elements would have the owned flag off, but the `<b>` and `<c>` elements would have the owned flag on.

The owned flag is set automatically. It is not part of the shape, and so does not affect `==`. Cloning an element creates a new element with the owned flag off. The owned flag is always off for an element that is immutable.

The `e.setChildren(xs)` works by doing the following to each constituent x of xs:

- if x is an element
    - if x has the owned flag on, let x = x.clone()
    - x now has the owned flag off
    - if x is mutable, take ownership of x, by setting the owned flag (do this atomically)
- if x is not an element, do nothing

Element construction does the same thing with children as setChildren.

The above construction process ensures that you can never construct an element with cycles.

There would be an additional function `xml:removeChildren(Element e)` that would give up ownership of the children. This would make `e` have empty children, and return the previous children of e after clearing the  owned flag.

**Comment by jclark** [comment]
Two things bother me about the approach of the previous comment

1. It raises the question of why elements cannot refer to their parents. One answer is that an immutable element can have arbitrarily many parents. It doesn't fit into immutability at all really. But that doesn't seem entirely satisfying.
2. It makes elements behave quite differently from other mutable structures.

**Comment by sanjiva** [comment]
While I see the 2nd point, I don't understand the first.

In our model elements don't refer to their parent, ever. Given that that is the model, what's the problem. I like the solution of "clone on adopt" basically for element child.

**Comment by jclark** [comment]
It weakens the justification for our model.

The norm in XML processing systems and APIs is that elements provide access to their parents. (I think this is partly because it's what people are used to, but partly because it is useful for some kinds of XML processing, but particularly text-oriented.) So I think we should have good reasons why we are not doing this.

The arguments for the way the current design does things (without the changes discussed in this issue) are:

1. Makes XML elements more like other mutable structures.
2. It allows XML elements to be shared between trees, which is useful in some integration scenarios
3. It works with our model of immutability.

If we do "clone on adopt", then that undermines points 1 and 2, leaving only 3, which feels like a weak justification.

**Comment by jclark** [comment]
I think having setChildren detect the cycle may be doable with adequate efficiency.  Basic idea is that elements maintain:

- list of weak references (java.lang.ref.WeakReference) to possible parents (typically this will be zero or one)
- integer level which is >=  1 + max of level of all element children

If you can assign such an integer level to every element, then you don't have a cycle.

So when you do a setChildren on an element, you see whether that increases the level. If so, you look at each possible parent and if it is still a parent, increase its level if necessary. If you did actually increase the level, then you recurse to their parent. If while recursing you hit an element you've already seen, you have a cycle.

(I only just thought of this, so it may well have fatal flaws.)

There's quite a lot of academic work on incremental graph cycle detection (which is effectively what this problem is). In particular:

https://www3.cs.stonybrook.edu/~bender/newpub/2015-BenderFiGi-SICOMP-topsort.pdf

I haven't studied this in detail, but it uses some similar concepts to the algorithm I sketched above. So even if the sketched algorithm doesn't work, I suspect there is something similar that will work.

**Comment by jclark** [comment]
As discussed today, initial implementation can just use list of weak references to possible parents. The levels concept is an optimization of that. Note in particular:

- checking a possible parent is an actual parent is potentially expensive, since it requires looking at the entire list of children (which may often be long); so it's better to check up the ancestry tree first (which is typically short), and if that reveals a potential cycle, only then check whether it is an actual parent and an actual cycle
- we can create a single Java WeakReference object shared by all element children of a given element

**Comment by rdhananjaya** [comment]
xml a = xml `<A></A>`;
xml b = xml `<B></B>`;
xml subWrap = xml `<W></W>`;
subWrap.setChildren(a);

a.setChildren(b);
b.setChildren(subWrap); // formed a cycle A -> B --insertion--> W -> A

To catch this form of cycles we do need to analyze the `child` tree after analyzing the parent list. The algorithm was modified to collect the actual `parents set` of the current node while analyzing the parent part of the tree and then if no cycle was detected, we perform a DFS on `child` tree, if we found any node in `parents set` in child tree that means we have a cycle.

**Comment by jclark** [comment]
I think there's some confusion here. Let's go through your example step by step:

```
xml:Element a = xml `<A></A>`; 
xml:Element b = xml `<B></B>`; 
xml:Element w = xml `<W></W>`;
w.setChildren(a); // 1
a.setChildren(b); // 2
b.setChildren(w); // 3
```

- after step 1, w has a ref to a, and a has a weak ref to w
- after step 2, a has a ref to b, and b has a weak ref to a
- if  step 3 completed, b has a ref to w, and w has a weak ref to b

Thus we would have cycle in the weak refs from w to b to a to w. So there's no need to search children.

**Comment by rdhananjaya** [comment]
Yes, I missed executing that logic in full. I removed the extra bits from my implementation and tests are detecting the loop just fine.
